### PR TITLE
Implemented generic REST lookup based auth

### PR DESCRIPTION
**Pull request by agoston**

### CODE REVIEW

File: src/server/mod.rs [hunk 1]
```diff
@@ -1015,6 +1007,18 @@ where
...
+                    let mut session = session.lock()?;
+                    session.state = WaitCmd;
+                    Ok(Reply::new(
+                        ReplyCode::UserLoggedIn,
+                        "User logged in, proceed",
+                    ))
+                }
+                Event::InternalMsg(AuthFailed) => Ok(Reply::new(
+                    ReplyCode::NotLoggedIn,
+                    "Wrong username or password",
```
Comment (hannesdejager) [inline src/server/mod.rs]: Perhaps just say Auth Failed. If an auth back-end that does say 2FA fails on the 2FA step then the username and password may be completely correct but the authentication still failed.

Comment (agoston) [reply src/server/mod.rs]: indeed, this far it was just a copy of the original error.

File: src/auth/pam.rs [hunk 2]
```diff
@@ -16,16 +19,35 @@ impl PAMAuthenticator {
...
-            Err(_) => Ok(false),
-        }
+                    auth.set_credentials(&username, &password);
+                    match auth.authenticate() {
+                        Ok(()) => Ok(true),
+                        Err(_) => Ok(false),
+                    }
+                })
+                .map_err(|_| {
+                    // FIXME: log error
```
Comment (hannesdejager) [inline src/auth/pam.rs]: Do you still intend to fix this, or is it coming later?

File: src/auth/rest.rs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        let username_json = encode_string_json(_username);
+        let password_json = encode_string_json(_password);
+        let body = self.fill_encoded_placeholders(&self.body, &username_json, &password_json);
+
+        // FIXME: need to clone too much, just to keep tokio::spawn() happy, with its 'static requirement. is there a way maybe to work this around with proper lifetime specifiers? Or is it better to just clone the whole object?
+        let method = self.method.clone();
+        let selector = self.selector.clone();
+        let regex = self.regex.clone();
+
+        //        println!("{} {}", url, body);
```
Comment (hannesdejager) [inline src/auth/rest.rs]: Pehaps we should just remove all the debugging by println :-). Consider replacing with `debug!`

File: src/auth/pam.rs (hunk 2 above)
Comment (hannesdejager) [reply src/auth/pam.rs]: I've replaced it with `debug!`

File: src/auth/rest.rs (hunk 3 above)
Comment (hannesdejager) [reply src/auth/rest.rs]: I've replaced it with `debug!`

**Pull request merged**